### ISSUE TITLE
Feature Request: Variables in validation reports of SPARQLConstraintComponent

### ISSUE THREAD

**Issue by baum314nk** [RDFLib/pySHACL#30]
I want to request a feature for validation reports of SPARQLConstraint(Component) as described in the [SHACL Recommondations](https://www.w3.org/TR/shacl/#sparql-constraints-variables).
I'd really appreciate the possibility to use variables from SELECT queries in the sh:message, i.e.:
```
:VerifyPowerAdapterSupplyShape
  a sh:NodeShape ;
  sh:targetClass ex:Computer ;
  sh:sparql [
    a sh:SPARQLConstraint ;
    sh:message "The power adapter ({?availablePower} W) must provide more power than the parts of the computer consume ({?requiredPower} W)." ;
    sh:prefixes ex: ;
    sh:select """
      SELECT $this ?availablePower ?requiredPower
      WHERE {
        $this ex:hasPowerAdapter ?powerAdapter .
        ?powerAdapter ex:hasPowerSupply ?availablePower .
        {
          SELECT (SUM(?power) as ?requiredPower)
          WHERE { 
	    $this ex:hasPart ?device .
	    ?device ex:hasRequiredPower ?power .
          }
        }
        FILTER(?availablePower < ?requiredPower) .
      }
    """ ;
  ] .
```

That would help me alot to debug ontologies that rely on complex SHACL-SPARQL constraints :)

**Comment by baum314nk** [comment]
I fixed this problem locally by modifying or adding the methods `_evaluate_sparql_constraint` & `_validate_sparql_query` of sparql_based_constraints.py and `make_v_result_description`, `make_v_result_description` & `_format_sparql_based_result_message` of constraint_component.py.

As I change the types of some parameters and return values, I'm not really sure that this workaround won't break some other things, but atm it works fine for me.
[pySHACL_constrains_quickfix.zip](https://github.com/RDFLib/pySHACL/files/3707349/pySHACL_constrains_quickfix.zip)

**Comment by ashleysommer** [comment]
Hi @baum314nk 
Thanks for opening this issue.
I had intended to implement this optional feature but never got around to it. There was always more important features to get working.

I will look at your patch to see if there is a way of easily integrating it into the rest of the codebase.
Thanks.

**Comment by piyush69** [comment]
Hi, I would like to work on this issue. I'll work backward from the patch and try to make it work with the existing design structure.

**Comment by ashleysommer** [comment]
@piyush69 
That would be great, if you could make as few changes to the existing code to reproduce the feature. Thanks for your help.

**Comment by piyush69** [comment]
@ashleysommer, I've put in a PR (#50) for the same. 
It makes the following changes:
 - add an optional function parameter `bound_vars` to some methods. It is `None` by default
 - add a function `_format_sparql_based_result_message()` in `sparql_based_constraints.py` to replace query variables with their corresponding values in the message

**Comment by ashleysommer** [comment]
@piyush69 
Thank you for your work on this.
I believe there is still one thing missing. I see it does inject the bound vars to the message in the descriptive output, but I think it also needs to inject the bound vars into the `sh:message` property in the validation report graph too.

**Comment by ashleysommer** [comment]
Forgot to close this. Closed by #50.
